### PR TITLE
Contextually type parameters in super calls using type arguments on the base class.

### LINKED ISSUE #1480: Overload resolution fails to use type arguments from base type in 'super' constructor calls
http://www.typescriptlang.org/Playground#src=class%20A%3Cx%2C%20y%3E%20%7B%0D%0A%09constructor(private%20map%3A%20(value%3A%20x)%20%3D%3E%20y)%20%7B%0D%0A%09%09%0D%0A%09%7D%0D%0A%7D%0D%0A%0D%0Aclass%20B%20extends%20A%3Cnumber%2C%20string%3E%20%7B%0D%0A%09%0D%0A%09constructor()%20%7B%20super(value%20%3D%3E%20String(value))%3B%20%7D%0D%0A%7D

image

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Seems like we might not even have been performing overload resolution correctly here.

Fixes #1480.

### CODE REVIEW

File: tests/baselines/reference/superCallArgsMustMatch.errors.txt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    
+    }
+    
+     
+    
+    class T6 extends T5<number>{
+    
+        constructor() {
+    
+            super("hi"); // Should error, base constructor has type T for first arg, which is fixed as number in the extends clause
```
Comment (JsonFreeman) [inline tests/baselines/reference/superCallArgsMustMatch.errors.txt]: Technically, the word you want is "instantiated" rather than "fixed"

File: tests/cases/fourslash/quickInfoForContextuallyTypedArrowFunctionInSuperCall.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+////
+////    }
+////}
+////
+////class B extends A<number, string> {
+////    constructor() { super(va/*1*/lue => String(va/*2*/lue.toExpone/*3*/ntial())); }
+////}
+
+goTo.marker('1');
+verify.quickInfoIs('(var) value: number');
```
Comment (JsonFreeman) [inline tests/cases/fourslash/quickInfoForContextuallyTypedArrowFunctionInSuperCall.ts]: Good test!

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -5978,11 +5978,36 @@ module ts {
...
+
         function resolveCall(node: CallLikeExpression, signatures: Signature[], candidatesOutArray: Signature[]): Signature {
             var isTaggedTemplate = node.kind === SyntaxKind.TaggedTemplateExpression;
 
-            var typeArguments = isTaggedTemplate ? undefined : (<CallExpression>node).typeArguments;
-            forEach(typeArguments, checkSourceElement);
+            var typeArguments: TypeNode[];
+
+            if (!isTaggedTemplate) {
+                typeArguments = getEffectiveTypeArguments(<CallExpression>node);
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: This is fine, but one alternative you might consider is for resolveCall to take a parameter for the type arguments, and have the caller determine what to pass. So whichever caller handles super calls would get the type arguments from the base type reference. But I admit it's a bit weird to pass the entire CallLikeExpression, with the the type arguments alongside it. It's like having meat and potatoes, with a side of potatoes.

File: tests/baselines/reference/superCallArgsMustMatch.errors.txt (hunk 1 above)
Comment (DanielRosenwasser) [inline tests/baselines/reference/superCallArgsMustMatch.errors.txt]: I didn't write it, but I'll consider fixing it.

File: src/compiler/checker.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Exactly! It's like if you gave me corn alongside corn-off-the-cob - it's kind of odd to give both.

**Comment by JsonFreeman** [comment]
:+1:

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -5978,11 +5978,36 @@ module ts {
             return args;
         }
 
+        // In a 'super' call, type arguments are not provided within the CallExpression node itself.
+        // Instead, they must be fetched from the class declaration's base type node.
+        function getEffectiveTypeArguments(callExpression: CallExpression): TypeNode[] {
+            if (callExpression.expression.kind === SyntaxKind.SuperKeyword) {
+                // TODO (drosen): 1) Discuss if checking needs to be done at this point.
```
Comment (yuit) [inline src/compiler/checker.ts]: I think #2 and #3 todo would be good to added in

Comment (JsonFreeman) [inline src/compiler/checker.ts]: +1

File: src/compiler/checker.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: We can do this without passing the type arguments. But I think we'll leave it for another PR.

**Comment by yuit** [comment]
lgtm
